### PR TITLE
feat(Guild): add api methods for membership screening

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**
Ref: discord/discord-api-docs#2396
A similar PR is coming to do the same with the Guild Welcome Screen

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/structures/Guild.js [hunk 1]
```diff
@@ -265,6 +267,12 @@ class Guild extends Base {
       this.embedChannelID = data.embed_channel_id;
     }
 
+    /**
+     * Whether Membership Screening is enabled on this guild
+     * @type {boolean}
+     */
+    this.membershipScreeningEnabled = this.features.includes('MEMBERSHIP_VERIFICATION_GATE_ENABLED');
```
Comment (NotSugden) [inline src/structures/Guild.js:274]: this should be a getter

File: src/structures/Guild.js [hunk 2]
```diff
@@ -959,6 +967,61 @@ class Guild extends Base {
       .then(data => GuildAuditLogs.build(this, data));
   }
 
+  /**
+   * Data for a field in Membership Screening
+   * @typedef {Object} GuildMembershipScreeningField
+   * @property {MembershipScreeningType} fieldType The type of the field
+   * @property {string} label The title of the field
+   * @property {string[]?} values The list of values in the field
```
Comment (NotSugden) [inline src/structures/Guild.js:975]:
```suggestion
   * @property {string[]} [values] The list of values in the field
```

File: src/structures/Guild.js [hunk 3]
```diff
@@ -959,6 +967,61 @@ class Guild extends Base {
...
+        fieldType: field.field_type,
+        label: field.label,
+        values: field.values,
+        required: field.required,
+      });
+    }
+    return {
+      enabled: this.membershipScreeningEnabled,
+      description: data.description,
+      formFields: formFields,
```
Comment (NotSugden) [inline src/structures/Guild.js:1021]: this could just be `formFields: data.form_fields.map(...);`

File: src/structures/Guild.js [hunk 4]
```diff
@@ -1422,6 +1485,54 @@ class Guild extends Base {
...
+    if (typeof memberScreen.formFields !== 'undefined') {
+      for (const field of memberScreen.formFields) {
+        form_fields.push({
+          field_type: field.field_type,
+          label: field.label,
+          values: field.values,
+          required: field.required,
+        });
+      }
+    }
```
Comment (NotSugden) [inline src/structures/Guild.js:1507]: once again, this can be perhaps
```js
const fields = memberScreen.formFields?.map(...);
```

File: src/structures/Guild.js [hunk 5]
```diff
@@ -1422,6 +1485,54 @@ class Guild extends Base {
...
+    let data = {};
```
Comment (NotSugden) [inline src/structures/Guild.js:1508]:
```suggestion
    const data = {};
```

File: src/structures/Guild.js [hunk 6]
```diff
@@ -1422,6 +1485,54 @@ class Guild extends Base {
...
+    if (typeof memberScreen.enabled !== 'undefined') data.enabled = memberScreen.enabled;
+    if (typeof memberScreen.description !== 'undefined') data.description = memberScreen.description;
+    if (typeof memberScreen.formFields !== 'undefined') data.form_fields = JSON.stringify(form_fields);
+    return this.client.api
+      .guilds(this.id, 'member-verification')
+      .patch({
+        data: data,
+      })
```
Comment (NotSugden) [inline src/structures/Guild.js:1516]:
```suggestion
      .patch({ data })
```

File: src/structures/Guild.js [hunk 7]
```diff
@@ -1422,6 +1485,54 @@ class Guild extends Base {
...
+      .then(res => {
```
Comment (NotSugden) [inline src/structures/Guild.js:1517]: could this use async/await rather than a `.then`

File: src/structures/Guild.js [hunk 8]
```diff
@@ -1422,6 +1485,54 @@ class Guild extends Base {
...
+            fieldType: field.field_type,
+            label: field.label,
+            values: field.values,
+            required: field.required,
+          });
+        }
+        return {
+          enabled: this.membershipScreeningEnabled,
+          description: res.description,
+          formFields: formFields,
```
Comment (NotSugden) [inline src/structures/Guild.js:1531]: seems this code is duplicated, in any case the same as before this can use the `map` function

File: typings/index.d.ts [hunk 9]
```diff
@@ -2664,6 +2670,25 @@ declare module 'discord.js' {
...
+    description: string;
+    enabled: boolean;
+    formFields: GuildMembershipScreeningField[];
+  }
+
+  interface GuildMembershipScreeningData {
+    description?: string;
+    enabled?: boolean;
+    formFields?: GuildMembershipScreeningField[];
+  }
```
Comment (NotSugden) [inline typings/index.d.ts:2683]:
```suggestion
  type GuildMembershipScreeningData = Partial<GuildMembershipScreening>;
```
either this or just directly use `Partial<GuildMembershipScreening>`

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: src/structures/Guild.js (hunk 8 above)
Comment (ckohen) [reply src/structures/Guild.js]: I noticed that there was a few instances of this necessary, would it be desirable to have a private function that transforms this data and call that in each location instead?

File: src/structures/Guild.js [hunk 10]
```diff
@@ -1422,6 +1476,38 @@ class Guild extends Base {
...
+    const fields = memberScreen.formFields?.map(field => ({
+      field_type: field.field_type,
+      label: field.label,
+      values: field.values,
+      required: field.required,
+    }));
+    const data = {};
+    if (typeof memberScreen.enabled !== 'undefined') data.enabled = memberScreen.enabled;
+    if (typeof memberScreen.description !== 'undefined') data.description = memberScreen.description;
+    if (typeof memberScreen.formFields !== 'undefined') data.form_fields = JSON.stringify(fields);
```
Comment (NotSugden) [inline src/structures/Guild.js:1497]:
```suggestion
    if (typeof memberScreen.formFields !== 'undefined') data.form_fields = fields;
```
i don't think this is meant to be stringified here

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: src/structures/Guild.js (hunk 10 above)
Comment (ckohen) [reply src/structures/Guild.js:1497]: In my testing (and according to the referenced docs) it has to be stringified. This may be an accident in the API, but this currently accurately reflects what the API is expecting.

Comment (NotSugden) [reply src/structures/Guild.js:1497]: oops, misread docs mb

File: src/structures/Guild.js [hunk 11]
```diff
@@ -177,6 +177,8 @@ class Guild extends Base {
      * * VERIFIED
      * * VIP_REGIONS
      * * WELCOME_SCREEN_ENABLED
+     * * MEMBER_VERIFICATION_GATE_ENABLED
+     * * PREVIEW_ENABLED
```
Comment (tipakA) [inline src/structures/Guild.js:181]: Shouldn't these be sorted alphabetically just like the rest was? Same goes for the typings file.

Comment (ckohen) [reply src/structures/Guild.js]: Guess I put them in the order discord docs have them in without noticing they're alphabetical here. Fixed!

File: src/structures/Guild.js [hunk 12]
```diff
@@ -1422,6 +1476,38 @@ class Guild extends Base {
       .then(() => this);
   }
 
+  /**
+   * Edits the guild's membership screening form.
+   * @param {Partial<GuildMembershipScreening>} memberScreen The membership screening data for the guild
```
Comment (NotSugden) [inline src/structures/Guild.js:1481]: could there be something that explains what `Partial<x>` is on the docs? 
image

**Review by NotSugden — CHANGES_REQUESTED** [review]

**Comment by almostSouji** [comment]
update: upstream  https://github.com/discord/discord-api-docs/pull/2396 has been merged and finalized

File: src/structures/Guild.js [hunk 13]
```diff
@@ -484,6 +486,15 @@ class Guild extends Base {
...
 
+  /**
+   * Whether Membership Screening is enabled on this guild
+   * @type {boolean}
+   * @readonly
+   */
+  get membershipScreeningEnabled() {
+    return this.features.includes('MEMBER_VERIFICATION_GATE_ENABLED');
+  }
+
```
Comment (advaith1) [inline src/structures/Guild.js:497]: this is not needed imo; .partnered and .verified were hacks on checking other features (added before the PARTNERED and VERIFIED features iirc) and should be removed too in a separate pr; `guild.membershipScreeningEnabled` is barely an improvement over `guild.features.has('MEMBER_VERIFICATION_GATE_ENABLED')`

File: src/util/Constants.js [hunk 14]
```diff
@@ -512,6 +512,13 @@ exports.ExplicitContentFilterLevels = ['DISABLED', 'MEMBERS_WITHOUT_ROLES', 'ALL
  */
 exports.VerificationLevels = ['NONE', 'LOW', 'MEDIUM', 'HIGH', 'VERY_HIGH'];
 
+/**
+ * The type of membership screening. Here are the available types:
+ * * TERMS
+ * @typedef {string} MembershipScreeningType
+ */
+exports.MembershipScreeningType = ['TERMS'];
```
Comment (advaith1) [inline src/util/Constants.js:520]: can we add the name here?

maybe
```suggestion
/**
 * The type of membership screening. Here are the available types:
 * * TERMS: Server Rules
 * @typedef {string} MembershipScreeningType
 */
exports.MembershipScreeningType = ['TERMS'];
```

File: src/structures/Guild.js [hunk 15]
```diff
@@ -959,6 +970,49 @@ class Guild extends Base {
...
+   * @example
+   * // Fetches the guild membership screening options
+   * guild.fetchMembershipScreening()
+   *   .then(memberScreen => console.log(`Membership Screening is ${memberScreen.enabled ? 'enabled' : 'disabled'}`))
+   *   .catch(console.error);
+   */
+  async fetchMembershipScreening() {
+    if (!this.features.includes('COMMUNITY')) {
+      throw new Error('COMMUNITY');
+    }
```
Comment (advaith1) [inline src/structures/Guild.js:1002]: I don't see this in https://github.com/discordjs/discord.js/blob/master/src/errors/Messages.js, and it doesn't seem to be added in this PR?

File: src/util/Constants.js (hunk 14 above)
Comment (ckohen) [reply src/util/Constants.js:520]: Good idea, i'll put `TERMS` in a code block though to make it a little clearer

File: src/structures/Guild.js (hunk 13 above)
Comment (ckohen) [reply src/structures/Guild.js:497]: yeah, i see your point, especially for v13 where we are trying to remove these simple wrappers, will remove

File: src/structures/Guild.js [hunk 16]
```diff
@@ -959,6 +970,49 @@ class Guild extends Base {
...
+
+  /**
+   * Fetches the guild Membership Screening data.
+   * @returns {Promise<GuildMembershipScreening>}
+   * @example
+   * // Fetches the guild membership screening options
+   * guild.fetchMembershipScreening()
+   *   .then(memberScreen => console.log(`Membership Screening is ${memberScreen.enabled ? 'enabled' : 'disabled'}`))
+   *   .catch(console.error);
+   */
```
Comment (advaith1) [inline src/structures/Guild.js:998]: maybe use an example that actually requires fetching it (can just check the feature for this, and users shouldn't be fetching it for this)

File: src/structures/Guild.js (hunk 15 above)
Comment (ckohen) [reply src/structures/Guild.js:1002]: Good point, I hadn't realized that Error's were extended when I wrote this, am aware now, will fix in next push

**Review by iCrawl — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by advaith1 — APPROVED** [review]

File: src/structures/Guild.js [hunk 17]
```diff
@@ -1422,6 +1470,45 @@ class Guild extends Base {
       .then(() => this);
   }
 
+  /**
+   * A `Partial` object is a representation of any existing object.
+   * This object contains between 1 and all of the original objects parameters.
```
Comment (SpaceEEC) [inline src/structures/Guild.js:1475]: Since you actually can edit with an empty object.
```suggestion
   * This object contains between 0 and all of the original objects parameters.
```

**Review by SpaceEEC — CHANGES_REQUESTED** [review]

File: src/structures/Guild.js [hunk 18]
```diff
@@ -959,6 +961,52 @@ class Guild extends Base {
...
+   *     const requiredFields = memberScreen.formFields.filter(field => field.required);
+   *     console.log(`Screening has ${memberScreen.formFields.length} steps, ${requiredFields.length} are required`);
+   *   })
+   *   .catch(console.error);
+   */
+  async fetchMembershipScreening() {
+    if (!this.features.includes('COMMUNITY')) {
+      throw new Error('COMMUNITY');
+    }
+    const data = await this.client.api.guilds(this.id)['member-verification'].get();
```
[the remark this answers is not in the document]
Comment (ckohen) [reply src/structures/Guild.js:997]: This may be a bug with how the API is currently handling this, see https://github.com/discord/discord-api-docs/issues/2530
A check is still a good idea though, will add.

**Comment by ckohen** [comment]
Looks like the associated API methods for this are being changed, I'll draft this PR until we get the updated ones.
Ref: https://github.com/discord/discord-api-docs/pull/2547

**Comment by iCrawl** [comment]
If this is still wanted (upstream PR has been merged) this should be undrafted and rebased.

**Comment by advaith1** [comment]
@iCrawl the docs were removed pending breaking changes, and Discord hasn't given any information since then

**Comment by iCrawl** [comment]
Right, I didn't check, I assumed that would be the "new" PR. Thats fine then.

File: src/structures/Guild.js (hunk 18 above)
Comment (NotSugden) [inline src/structures/Guild.js:997]:
```suggestion
    const data = await this.client.api.guilds(this.id, 'member-verification').get();
```

Comment (ckohen) [reply src/structures/Guild.js:997]: I'll address this when I can resume work on this PR.
